### PR TITLE
Add backpressure to log monitor

### LINKED ISSUE #7317: ray logging_monitor steadily increases CPU utilization
### What is the problem?
Over the course of six days I've observed the ray logging_monitor steadily increase the percent of CPU consumed. The logging_monitor corresponds to Ray worker running in --head mode with two Ray drivers connecting to it--one to submit jobs to Ray and the other pulling completed jobs from the underlying Plasma Object Store. Possibly important note: the driver connections remain open over multiple days.

*Ray version and other system information (Python version, TensorFlow version, OS):*
Ray: 0.7.5
OS: Ubuntu 16.04

### Reproduction (REQUIRED)
Run a Ray worker in head mode for 3 or more days straight, submitting jobs and removing the resulting objects from Ray.

After 5 days top in the Docker container (pid 36):

ray-container-ps-aux

ray-log_monitor_5_days

At 6 days:

ray-log_monitor-6-days

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

Rn, log monitor can use excessive cpu usage because ray cluster can have many files in the log directory (1000+) when it runs for a long time. Based on the profiling result, the root cause is calling glob 5~6 times in "update log files" to 1000+ files. There are 2 approaches;

- Blacklist closed files before calling glob.
- Add backpressure in the loop. Right now, it doesn't sleep if nothing is published, and it sleeps 0.1 second if anything is published.

This PR addresses the second case since that's easy. 

## Related issue number

Half of https://github.com/ray-project/ray/issues/17675
closes https://github.com/ray-project/ray/issues/7317

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by ericl — CHANGES_REQUESTED** [review]
I don't think this is a good idea. It's using 100% CPU because there are a lot of log files; which is fine. What this PR does is slows down log collection as you add more files, which is exactly the opposite of what you want.

How about optimizing this instead?

**Comment by rkooo567** [comment]
Screen Shot 2021-08-12 at 1 53 19 PM

Actually if you look at this user flame graph, most of overhead is from globbing files in the while loop (which happens in `update_log_filenames`).

I think reading log file itself has smaller overhead because we remember how much we read so far (`check_log_files_and_publish_updates`).

But I understand your concern (and that can totally happen). If we can prioritize this task as an usability issue (since there's an user suffering from it), I am down to do optimization with more uni tests in this layer. 

Or I can try asking a blocked user if this fixes their issue with the custom wheel and work on optimization later.

**Comment by ericl** [comment]
Can you put this in the backlog as a P1 bug?

File: python/ray/_private/log_monitor.py [hunk 1]
```diff
@@ -140,6 +140,7 @@ def update_log_filenames(self):
         # runtime_env setup process is logged here
         runtime_env_setup_paths = glob.glob(
```
Comment (ericl) [inline python/ray/_private/log_monitor.py:141]: Another thing we can do is update the list of files scanned more infrequently.

**Comment by rkooo567** [comment]
Sure. I think high memory from the dashboard is probably more critical bug. But I bumped up the priority and I will close this PR.

File: python/ray/_private/log_monitor.py (hunk 1 above)
Comment (rkooo567) [reply python/ray/_private/log_monitor.py:141]: Oh yeah actually that sounds good. I think we can just do that in this pr?

Comment (ericl) [reply python/ray/_private/log_monitor.py:141]: Sure, it sounds good.

Comment (rkooo567) [reply python/ray/_private/log_monitor.py:153]: Update every 0.5 second if there are more than 1000 files now.

File: python/ray/_private/log_monitor.py [hunk 2]
```diff
@@ -297,8 +312,18 @@ def run(self):
         This will query Redis once every second to check if there are new log
         files to monitor. It will also store those log files in Redis.
         """
+        total_log_files = 0
+        last_updated = datetime.now()
         while True:
-            self.update_log_filenames()
+            elapsed_seconds = (datetime.now() - last_updated).seconds
+            logger.error(f"files: {self.log_filenames}")
```
Comment (ericl) [inline python/ray/_private/log_monitor.py:319]: Remove error

File: python/ray/_private/log_monitor.py [hunk 3]
```diff
@@ -297,8 +312,18 @@ def run(self):
         This will query Redis once every second to check if there are new log
         files to monitor. It will also store those log files in Redis.
         """
+        total_log_files = 0
+        last_updated = datetime.now()
```
Comment (ericl) [inline python/ray/_private/log_monitor.py:316]: time.time()?

File: python/ray/_private/log_monitor.py [hunk 4]
```diff
@@ -297,8 +312,18 @@ def run(self):
...
         while True:
-            self.update_log_filenames()
+            elapsed_seconds = (datetime.now() - last_updated).seconds
+            logger.error(f"files: {self.log_filenames}")
+            if (total_log_files < RAY_LOG_MONITOR_MANY_FILES_THRESHOLD
+                    or elapsed_seconds > LOG_NAME_UPDATE_INTERVAL_S):
+                logger.error(f"total log files: {total_log_files}")
+                logger.error(f"time elapsed: {elapsed_seconds}")
```
Comment (ericl) [inline python/ray/_private/log_monitor.py:323]: stray error logs

**Review by ericl — APPROVED** [review]
LGTM (clean up logs?)

**Pull request merged**